README.md
# GANETI CONTROL CENTER

## What is this?

GCC is a frontend to manage ganeti clusters. Essentially, it is a web application based on Slim/Twig, that wraps around a PHP class ```ganetiClient```, which in turn connects to a cluster's RAPI daemon. The class could also be used in other projects/scripts. As of now, GCC does not store any data on its own. However, you need a TFTP server running on the same machine to use the 'preseed feature' (which starts an instance and creates a special tftp/pxe boot configuration for this host).

## Who should use this and who should not?

GCC is tightly build around existing infrastructure, therefore it makes certain assumptions or at least works best in this environment:
* Hypervisor: KVM
* Disk Templates: DRBD
* Instances: independent virtual machines with partition tables, a bootloader etc. (no deboostrap, no integration with ganeti whatsoever)
* Network: the nodes are connected to multiple instance VLANs, each with its own bridge interface namend ```br<VLAN``` on the node - instance interfaces are connected to these bridges
* Remote Console: Spice

## Features

* supports multiple ganeti clusters
* list all instances
* show instance details
* manage instances (create, start, shutdown, migrate)
* spice console integration (tested only on linux)
* list cluster nodes/groups
* list/manipulate cluster settings
* list jobs
* show cluster statistics

## Ganeti Setup

To enable the RAPI daemon on a Debian based system, please update your ```/etc/default/ganeti``` file to contain the following line:
```
RAPI_ARGS="--require-authentication"
```
You also need to create a user/password pair, which is well covered in the official [RAPI Documentation](http://docs.ganeti.org/ganeti/2.15/html/rapi.html#users-and-passwords).

The RAPI daemon does only run the master node - if your cluster master IP is reachable you can point GCC to this address. However, if your cluster uses a separate network for cluster communication the master's IP address might not be reachable from the outside. As of now, the only possible solution to this is to specify the current master in the GCC configuration and update it, when the master changes. In the future, GCC might figure out the available cluster nodes itself (once it has a working connection) and switch to a new master automagically. 

## Dependencies

* apache + mod-rewrite
* php5
* php5-curl
* TFTP Server

## GCC Setup

* Install dependencies 
* this repository
* rename ```config/config.inc.php.dist``` to ```config/config.inc.php``` and edit
* the webserver needs write access to the folder ```/var/lib/tftpboot/pxelinux.cfg```
* point your webserver's root to ```ganeti-control-center/public```
* configure HTTPS + LDAP auth in your webserver

## missing features / ideas / TODO

* destructive actions (e.g. instance removal)
* store all cluster nodes locally and remember the current master - switch automatically if it becomes unreachable
* DHCP integration: allocate an IP address for a new instance or at least figure out which IP address has been assigned by the DHCP server
* DNS integration: create DNS records via nsupdate? 

### Screenshots

[<img src="http://s15.postimg.org/mms00ql93/gcc_clusternodes.jpg">](http://s15.postimg.org/i0vvsdzq3/gcc_clusternodes.png)
[<img src="http://s8.postimg.org/98xr6m9g1/gcc_instancedetails.jpg">](http://s8.postimg.org/qm81lh4r9/gcc_instancedetails.png)
[<img src="http://s11.postimg.org/wvg91a467/gcc_instances.jpg">](http://s11.postimg.org/byk0wm65f/gcc_instances.png)
[<img src="http://s32.postimg.org/rdogtuh8h/gcc_stats.jpg">](http://s32.postimg.org/apwyrcmh1/gcc_stats.png)


debian/changelog
ganeti-control-center (0.1) unstable; urgency=low

  * Initial debian package release

 -- Rudolph Bott <bott@sipgate.de>  Wed, 14 Mar 2018 08:38:12 +0100


public/assets/js/gcc.js
function isNumber(n) {
	return /^-?[\d.]+(?:e-?\d+)?$/.test(n);
}

function bytesToSize(bytes) {
	var sizes = ['MB', 'GB', 'TB'];
	if (bytes == 0) return 'n/a';
	var i = parseInt(Math.floor(Math.log(bytes) / Math.log(1024)));
	return Math.round(bytes / Math.pow(1024, i), 2) + ' ' + sizes[[i]];
}

function updateJobStatusModal() {
	var jobId = $( "#jobStatusModal" ).data( "jobId" );
	$.getJSON("/jobStatus/" + jobId , function(data) {
		$( "#jobStatusModal").find( ".modal-body" ).html( "<p>Current Job State: <span class=\"job_" + data + "\">" + data + "</span><img class=\"pull-right\" id=\"jobStatusLoader\" src=\"/assets/img/ajax-loader.gif\"></p>" );
		if (data != "success" && data != "error") {
			setTimeout(updateJobStatusModal, 1000);
		}
		else {
			if (data == "success") {	
				setTimeout(function() {
					$( "#jobStatusModal").modal('hide');
					location.reload();
				}, 1000);
			}
			else {
				$( "#jobStatusLoader" ).remove();
			}
		}
	});
}

function processResult(data) {
	if( isNumber(data) && data > 0) {
		$( "#jobStatusModal" ).find( ".modal-title" ).text( "Job Status for #" + data );
		$( "#jobStatusModal" ).data( "jobId", data);
		$( "#jobStatusModal").find( ".modal-body" ).html( "<p><img class=\"pull-right\" id=\"jobStatusLoader\" src=\"/static/img/ajax-loader.gif\"></p>" );
		setTimeout(updateJobStatusModal, 1000);
		$( "#jobStatusModal" ).modal();
	}
	else {
		$( "#jobStatusModal").find( ".modal-title" ).text( "Error" );
		$( "#jobStatusModal").data( "jobId", '-1');
		if(data.message) {
			var errorMsg = "Error-Code: " + data.code + "<br>\n" + "Message: " + data.message;
		}
		else {
			var errorMsg = "The request failed (unknown reason)";
		}
		$( "#jobStatusModal").find( ".modal-body" ).html( "<div class=\"alert alert-danger\" role=\"alert\">" + errorMsg + "</div>" );
		$( "#jobStatusModal").modal();
	}
}

function setInstanceParameter(instance, type, param, value) {
	$.post ( "/instanceParameter/" + instance + "/" + type + "/" + param + "/" + value, null, processResult, 'json');
}

function changeInstanceStatus(instance, instanceAction) {
	$.post ( "/changeInstanceStatus/" + instance, { 'action': instanceAction }, processResult, 'json')
}

function updateNic(instance, nicId, mac, link) {
	$.post ( "/updateNic/" + instance + "/" + nicId + "/" + mac + "/" + link, null, processResult, 'json')
}

function addNic(instance, mac, link) {
	$.post ( "/addNic/" + instance + "/" + mac + "/" + link, null, processResult, 'json')
}

function growDisk(instance, diskId, amount) {
	$.post ( "/growDisk/" + instance + "/" + diskId + "/" + amount, null, processResult, 'json')
}

function addDisk(instance, size) {
	$.post ( "/addDisk/" + instance + "/" + size, null, processResult, 'json')
}

function addTag(instance, tag) {
	$.post ( "/addTag/" + instance + "/" + tag, null, processResult, 'json')
}

function setClusterHvParameter(type, param, value) {
	$.post ( "/clusterHvParameter/" + type + "/" + param + "/" + value, null, processResult, 'json');
}

function setClusterIpolicyParameter(param, value) {
	$.post ( "/clusterIpolicyParameter/" + param + "/" + value, null, processResult, 'json');
}

$( document ).ready(function() {
	$( ".paramChange" ).click(function() {
		var instance = $(this).data("instance");
		var type = $(this).data("type");
		var param = $(this).data("param");
		var value = $(this).data("value");
		setInstanceParameter(instance, type, param, value);
	});

	$( ".instanceAction" ).click(function() {
		var instance = $(this).data("instance");
		var instanceAction = $(this).data("instanceaction");
		$.confirm({
			text: "Are you sure you want to perform the action '" + instanceAction + "' on Instance '" + instance + "'?",
			confirm: function() { changeInstanceStatus(instance, instanceAction); },
		});

	});
});


includes/ganetiClient.php
<?php

# ganeti-control-center
# Copyright (C) 2016 sipgate GmbH
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

class ganetiClient {
	private $config;

	function ganetiClient($config) {
		$this->config = $config;
	}

	function callApi($method, $url, $data = false)
	{
		$curl = curl_init();

		switch ($method)
		{
		case "POST":
			curl_setopt($curl, CURLOPT_POST, 1);

			if ($data) {
				$data_string = json_encode($data);
				curl_setopt($curl, CURLOPT_HTTPHEADER, array(                                                                          
					'Content-Type: application/json',                                                                                
					'Content-Length: ' . strlen($data_string))                                                                       
				);

				curl_setopt($curl, CURLOPT_POSTFIELDS, $data_string);
			}
			break;
		case "PUT":
			curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "PUT");

			if ($data) {
				$data_string = json_encode($data);
				curl_setopt($curl, CURLOPT_HTTPHEADER, array(                                                                          
					'Content-Type: application/json',                                                                                
					'Content-Length: ' . strlen($data_string))                                                                       
				);

				curl_setopt($curl, CURLOPT_POSTFIELDS, $data_string);
			}

            break;
        case "PUT_W_GET_PARAMS":
            # this is only here to make the broken "tags" part of RAPI work
            curl_setopt($curl, CURLOPT_CUSTOMREQUEST, "PUT");

			if ($data)
				$url = sprintf("%s?%s", $url, http_build_query($data));

            break;
		default:
			if ($data)
				$url = sprintf("%s?%s", $url, http_build_query($data));
		}

		// Optional Authentication:
		curl_setopt($curl, CURLOPT_HTTPAUTH, CURLAUTH_BASIC);
		curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, false);
		curl_setopt($curl, CURLOPT_SSL_VERIFYHOST, false);
		curl_setopt($curl, CURLOPT_USERPWD, $this->config["user"] . ":" . $this->config["password"]);

		curl_setopt($curl, CURLOPT_URL, "https://" . $this->config["host"] . ":5080" . $url);
		curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);

		$result = curl_exec($curl);
		print_r(curl_error($curl));

		curl_close($curl);

		return $result;
	}

	function getConfigName() {
		return $this->config["name"];
	}

	function getClusterInfo() {
		$data = json_decode($this->callApi("GET","/2/info"),true);
		return $data;
	}

	function getNodes($bulk = false) {
		if($bulk) {
			$data = json_decode($this->callApi("GET","/2/nodes",array("bulk" => "1")),true);
		}
		else {
			$data = json_decode($this->callApi("GET","/2/nodes"),true);	
		}
		return $data;
	}

	function getInstances($bulk = false) {
		if($bulk) {
			$data = json_decode($this->callApi("GET","/2/instances",array("bulk" => "1")),true);
		}
		else {
			$data = json_decode($this->callApi("GET","/2/instances"),true);
		}
		return $data;
	}

	function getInstance($name) {
		$data = json_decode($this->callApi("GET","/2/instances/" . $name),true);
		return $data;
	}

	function getInstanceConsole($name) {
		$data = json_decode($this->callApi("GET","/2/instances/" . $name . "/console"),true);
		return $data;
	}

	function createInstance($params) {
		$data = json_decode($this->callApi("POST","/2/instances", $params),true);
		return $data;
    }

    function createMultiInstance($params) {
        $data = json_decode($this->callApi("POST","/2/instances-multi-alloc", $params),true);
        return $data;
    }

	function startInstance($instance) {
		$data = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/startup"),true);
		return $data;
	}

	function shutdownInstance($instance) {
		$data = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/shutdown"),true);
		return $data;
	}

	function failoverInstance($instance) {
		$data = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/failover"),true);
		return $data;
	}

	function migrateInstance($instance) {
		$data = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/migrate"),true);
		return $data;
	}

	function setInstanceParameter($instance, $type, $parameter, $value) {
		switch($type) {
		case "beparam":
			$data = array(
				"beparams" => array(
					$parameter => $value),
			);
			$return = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/modify",$data), true);
			return $return;
			break;
		case "hvparam":
			$data = array(
				"hvparams" => array(
					$parameter => $value),
			);
			$return = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/modify", $data), true);
			return $return;
			break;
		}
	}

	function setInstanceParameters($instance, $beparams = array(), $hvparams = array()) {
		if(!empty($beparams)) $data["beparams"] = $beparams;
		if(!empty($hvparams)) $data["hvparams"] = $hvparams;
		$return = json_decode($this->callApi("PUT","/2/instances/" . $instance . "/modify", $data), true);
		return $return;
	}

	function setClusterParameter($section, $type, $parameter, $value) {
		$return = -1;
		switch($section) {
		case "hvparams":
			$data = array(
				"hvparams" => array(
					$type => array(
						$parameter => $value,
					),
				),
			);
			$return = json_decode($this->callApi("PUT", "/2/modify", $data), true);
			break;
		case "ipolicy":
			$data = array(
				"ipolicy" => array(
					$parameter => $value
				),
			);
			$return = json_decode($this->callApi("PUT", "/2/modify", $data), true);
		}
		return $return;
	}

	function setNicParameters($instance, $nic, $mac, $link) {
		$data = array(
			"nics" => array(
				array(
					"modify",
					$nic,
					array(
						"link" => $link,
					),
				),
			),
		);

		$origInstance = $this->getInstance($instance);
		if($origInstance["nic.macs"][$nic] != $mac) {
			$data["nics"][0][2]["mac"] = $mac;
		}

		return json_decode($this->callApi("PUT","/2/instances/" . $instance . "/modify", $data), true);
	}

    function addNic($instance, $mac, $link) {
		$data = array(
			"nics" => array(
				array(
					"add",
					-1,
                    array(
                        "mac" => $mac,
						"link" => $link,
					),
				),
			),
		);

		$origInstance = $this->getInstance($instance);

		return json_decode($this->callApi("PUT","/2/instances/" . $instance . "/modify", $data), true);
    }

    function addDisk($instance, $size) {
		$data = array(
			"disks" => array(
				array(
					"add",
					-1,
                    array(
                        "size" => $size . "g"
					),
				),
			),
		);

		$origInstance = $this->getInstance($instance);

		return json_decode($this->callApi("PUT","/2/instances/" . $instance . "/modify", $data), true);
    }

    function addTag($instance, $tag) {
        $data["tag"] = $tag;
        return json_decode($this->callApi("PUT_W_GET_PARAMS","/2/instances/" . $instance . "/tags", $data), true);
    }

    function growDisk($instance, $disk, $amount) {
        $data = array(
            "amount" => $amount,
            "absolute" => false,
            "wait_for_sync" => false,
        );

		return json_decode($this->callApi("POST","/2/instances/" . $instance . "/disk/" . $disk . "/grow", $data), true);
    }

	function getJobs($bulk = false) {
		if($bulk) {
			$data = json_decode($this->callApi("GET","/2/jobs",array("bulk" => "1")),true);
			for($i = 0; $i < count($data); $i++) {
				if(isset($data[$i]["summary"][0]) && preg_match('/^INSTANCE_.+\((.+)\)$/',$data[$i]["summary"][0],$hits)) {
					$data[$i]["instanceLink"] = $hits[1];
				}
				else {
					$data[$i]["instanceLink"] = false;
				}
			}
		}
		else {
			$data = json_decode($this->callApi("GET","/2/jobs"),true);
		}
		return $data;
	}

	function getJob($jobid) {
		$data = json_decode($this->callApi("GET","/2/jobs/" . $jobid),true);
		return $data;
	}

	function getGroups($bulk = false) {
		if($bulk) {
			$data = json_decode($this->callApi("GET","/2/groups",array("bulk" => "1")),true);
		}
		else {
			$data = json_decode($this->callApi("GET","/2/groups"),true);
		}
		return $data;
	}

	function getStats() {
		$stats["instance-count"] = 0;
		$stats["instances"]["memory"] = 0;
		$stats["instances"]["disk"] = 0;
		$stats["instances"]["cpus"] = 0;
		$stats["node-count"] = 0;
		$stats["nodes"]["memory"] = 0;
		$stats["nodes"]["disk"] = 0;
		$stats["nodes"]["cpus"] = 0;


		# load instances
		$instances = $this->getInstances(true);
		foreach($instances AS $instance) {
			# sum up memory/cpus (cluster-wide)
			$stats["instance-count"]++;
			$stats["instances"]["memory"] += $instance["beparams"]["memory"];
			$stats["instances"]["cpus"] += $instance["beparams"]["vcpus"];

			# sum up memory/cpus (per node)
			isset($stats["pernode"][$instance["pnode"]]["memory"]) ? $stats["pernode"][$instance["pnode"]]["memory"] += $instance["beparams"]["memory"] : $stats["pernode"][$instance["pnode"]]["memory"] = $instance["beparams"]["memory"];
				isset($stats["pernode"][$instance["pnode"]]["vcpus"]) ? $stats["pernode"][$instance["pnode"]]["vcpus"] += $instance["beparams"]["vcpus"] : $stats["pernode"][$instance["pnode"]]["vcpus"] = $instance["beparams"]["vcpus"];

				# same for any disk(s) that might exist
				foreach($instance["disk.sizes"] AS $disk) {
					$stats["instances"]["disk"] += $disk;
					isset($stats["pernode"][$instance["pnode"]]["disk"]) ? $stats["pernode"][$instance["pnode"]]["disk"] += $disk : $stats["pernode"][$instance["pnode"]]["disk"] = $disk;
				}

			# count instances per node
			isset($stats["pnode-counter"][$instance["pnode"]]) ? $stats["pnode-counter"][$instance["pnode"]]++ : $stats["pnode-counter"][$instance["pnode"]] = 1;
		}
		# calculate instances-per-node-data
		$stats["min-inst-per-node"] = min($stats["pnode-counter"]);
		$stats["max-inst-per-node"] = max($stats["pnode-counter"]);
		$stats["avg-inst-per-node"] = round(array_sum($stats["pnode-counter"]) / count($stats["pnode-counter"]));
		unset($stats["pnode-counter"]);

		# load cluster nodes
		$nodes = $this->getNodes(true);
		foreach($nodes AS $node) {
			# sum up memory/cpu/disk totals (cluster-wide)
			$stats["node-count"]++;
			$stats["nodes"]["memory"] += $node["mtotal"];
			$stats["nodes"]["disk"] += $node["dtotal"];
			$stats["nodes"]["cpus"] += $node["ctotal"];

			# save per-node data
			$stats["pernode"][$node["name"]]["memtotal"] = $node["mtotal"];
			$stats["pernode"][$node["name"]]["disktotal"] = $node["dtotal"];
			$stats["pernode"][$node["name"]]["cpus"] = $node["ctotal"];
		}

		# calculate memory/disk diffs
		$stats["nodes"]["memfree"] = $stats["nodes"]["memory"] - $stats["instances"]["memory"];
		$stats["nodes"]["diskfree"] = $stats["nodes"]["disk"] - $stats["instances"]["disk"];
		foreach($stats["pernode"] AS $node => $data) {
			@$stats["pernode"][$node]["memfree"] = $data["memtotal"] - $data["memory"];
			@$stats["pernode"][$node]["diskfree"] = $data["disktotal"] - $data["disk"];
		}

		# calculate real-to-virtual-cpu-ratio
		$stats["nodes"]["cpu-ratio"] = round($stats["instances"]["cpus"] / $stats["nodes"]["cpus"],2);

		return $stats;
	}
 
}
?>


public/index.php
<?php

# ganeti-control-center
# Copyright (C) 2016 sipgate GmbH
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

require_once('../vendor/autoload.php');

require_once('/etc/ganeti-control-center/config.inc.php');
require_once('../includes/ganetiClient.php');

\Slim\Slim::registerAutoloader();

$app = new \Slim\Slim(array(
	'templates.path' => '../tpl'
));

$app->view(new \Slim\Views\Twig());
$app->view->parserExtensions = array(new \Slim\Views\TwigExtension());

$clusterIndex = $app->getCookie('clusterIndex');
if(!isset($clusterIndex) || !isset($config["rapi"][$clusterIndex])) {
	$clusterIndex = 0;
	$app->setCookie('clusterIndex', '0', '365 days');
}
$config["rapi-current"] = $config["rapi"][$clusterIndex];


$app->get('/', function() use ($app) {
	$app->redirect("/instances");
});

$app->get('/instances(/:filter(/:value))', function($filter = "none", $value = "") use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$cluster = $g->getClusterInfo();
	$clusterName = $g->getConfigName();
	$instances = $g->getInstances(true);
	$instancesFiltered = array();
	switch($filter) {
	case "byPrimaryNode":
		for($i = 0; $i < count($instances); $i++) {
			if($instances[$i]["pnode"] == $value) {
				$instancesFiltered[] = $instances[$i];
			}
		}
		$instances = $instancesFiltered;
		break;
	case "bySecondaryNode":
		for($i = 0; $i < count($instances); $i++) {
			if($instances[$i]["snodes"][0] == $value) {
				$instancesFiltered[] = $instances[$i];
			}
		}
		$instances = $instancesFiltered;
		break;
	}
	$app->render('page_instances.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"cluster" => $cluster,
		"instances" => $instances,
		));
});

$app->get('/instanceDetails/:h', function($instanceName) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$clusterName = $g->getConfigName();
	$instance = $g->getInstance($instanceName);
	$pageTitle = $instanceName;
	if(isset($instance["nic.macs"][0])) {
		$mac = $instance["nic.macs"][0];
		$filename = "01-" . str_replace(":","-",$mac);
		$preseedConfigExists = file_exists("/var/lib/tftpboot/pxelinux.cfg/" . $filename);
	}
	else {
		$preseedConfigExists = false;
	}
	$app->render('page_instancedetails.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"pageTitle" => $pageTitle,
		"instance" => $instance,
		"instance_dump" => print_r($instance,true),
		"preseedConfigExists" => $preseedConfigExists,
		"vlans" => $config["vlans"][$clusterName],
		"spiceCa" => $config["rapi-current"]["spice-ca"],
		));
});

$app->get('/clusterNodes', function() use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$cluster = $g->getClusterInfo();
	$clusterName = $g->getConfigName();
	$nodes = $g->getNodes(true);
	$groups = $g->getGroups(true);
	for($i = 0; $i < count($nodes); $i++) {
		if(!empty($nodes[$i]["group.uuid"])) {
			foreach($groups AS $group) {
				if($group["uuid"] == $nodes[$i]["group.uuid"]) {
					$nodes[$i]["group.name"] = $group["name"];
					$nodes[$i]["group.policy"] = $group["alloc_policy"];
					break;
				}
			}
		}
	}
	$app->render('page_clusternodes.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"cluster" => $cluster,
		"nodes" => $nodes,
		"groups" => $groups,
		));
});

$app->get('/jobs', function() use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$cluster = $g->getClusterInfo();
	$clusterName = $g->getConfigName();
	$jobs = array_reverse($g->getJobs(true));
	$app->render('page_jobs.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"cluster" => $cluster,
		"jobs" => $jobs));
});

$app->get('/jobDetails/:h', function($jobid) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$clusterName = $g->getConfigName();
	$job = $g->getJob($jobid);
	$app->render('page_jobdetails.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"job" => $job,
		"job_dump" => print_r($job,true)));
});

$app->get('/jobStatus', function() use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$clusterName = $g->getConfigName();
	$jobs = array_reverse($g->getJobs(true));
	$status = array();
	foreach($jobs AS $job) {
		$status[$job["id"]] = $job["status"];
	}
	Header("Content-Type: application/json");
	echo json_encode($status);
});

$app->get('/jobStatus/:j', function($jobId) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$clusterName = $g->getConfigName();
	$job = $g->getJob($jobId);
	$status = $job["status"];
	Header("Content-Type: application/json");
	echo json_encode($status);
});

$app->get('/createInstance', function() use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$cluster = $g->getClusterInfo();
	$clusterName = $g->getConfigName();
	$app->render('page_createinstance.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"vlans" => $config["vlans"][$clusterName],
		"cluster" => $cluster,
		));
});

$app->post('/createInstance', function() use ($app) {
	global $config;
    $g = new ganetiClient($config["rapi-current"]);

	$params = array(
        "iallocator" => "hail",
    );
    $instances = array();
    foreach($_POST["vmName"] AS $instanceName) {
	    $instances[] = array(
			"conflicts_check" => false,
			"disk_template" => "drbd",
			"ip_check" => false,
			"mode" => "create",
			"name_check" => false,
			"no_install" => true,
			"wait_for_sync" => false,
			"instance_name" => $instanceName,
            "os_type" => "debootstrap+default",
            "hypervisor" => "kvm",
			"beparams" => array(
				"vcpus" => $_POST["cpuCount"],
				"memory" => $_POST["memory"],
			),
			"nics" => array(
				array(
					"link" => "br" . $_POST["vlan"],
					"mac" => "generate"
				),
			),
			"disks" => array(
				array(
					"size" => ($_POST["diskSpace"]*1024)
				),
			),
			"hvparams" => array(
				"kernel_path" => "",
				"machine_version" => "pc",
				"nic_type" => "paravirtual"
            ),
        );
        if(!empty($_POST["instanceTag"])) {
            $key = count($instances) - 1;
            $instances[$key]["tags"] = array($_POST["instanceTag"]);
        }
    }
    $params["instances"] = $instances;
    $g->createMultiInstance($params);
	$app->redirect("/jobs");
});

$app->post('/changeInstanceStatus/:h', function($instance) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = -1;
	switch($_POST["action"]) {
	case "start":
		$return = $g->startInstance($instance);
		break;
	case "shutdown":
		$return = $g->shutdownInstance($instance);
		break;
	case "migrate":
		$return = $g->migrateInstance($instance);
		break;
	case "preseed":
		$inst = $g->getInstance($instance);
		if(!$inst["oper_state"]) {
			$mac = $inst["nic.macs"][0];
			$filename = "01-" . str_replace(":","-",$mac);
			if(copy("/etc/ganeti-control-center/pxe-boot-template.txt","/var/lib/tftpboot/pxelinux.cfg/" . $filename)) {
				$return = $g->startInstance($instance);
			}
		}
		break;
	}
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->get('/removePreseedConfig/:h', function($instance) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$inst = $g->getInstance($instance);
	$mac = $inst["nic.macs"][0];
	$filename = "01-" . str_replace(":","-",$mac);
	if(file_exists("/var/lib/tftpboot/pxelinux.cfg/" . $filename)) {
		unlink("/var/lib/tftpboot/pxelinux.cfg/" . $filename);
	}
	$app->redirect("/instanceDetails/" . $instance);
});

$app->get('/setCluster/:h', function($cluster) use ($app) {
	global $config;
	if(is_numeric($cluster) && isset($config["rapi"][$cluster])) {
		$app->setCookie("clusterIndex", $cluster, "365 days");
	}
	else {
		$app->setCookie("clusterIndex", "0", "365 days");
	}
	$app->redirect("/instances");
	exit;
});

$app->get('/getSpiceConfig/:i', function($instance) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$inst = $g->getInstance($instance);
	
	Header("Content-Type: application/x-virt-viewer");
	header("Content-disposition: attachment; filename=spice-connection.ini");
	echo "[virt-viewer]\n";
        echo "type=spice\n";
	echo "host=" . $inst["pnode"] . "\n";
	if($inst["hvparams"]["spice_use_tls"]) {
		echo "tls-port=" . $inst["network_port"] . "\n";
		echo "host-subject=CN=ganeti.example.com\n";
	}
	else {
		echo "port=" . $inst["network_port"] . "\n";
	}
	echo "delete-this-file=1\n";
	exit;
});

$app->post('/instanceParameter/:h/:t/:p/:v', function($instance, $type, $parameter, $value) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = -1;
	switch($type) {
	case "beparam":
		switch($parameter) {
		case "autoBalance":
			switch($value) {
			case "0":
				$return = $g->setInstanceParameter($instance, $type, "auto_balance", false);
				break;
			case "1":
				$return = $g->setInstanceParameter($instance, $type, "auto_balance", true);
				break;
			}
			break;
		case "spindleUseCount":
			if(is_numeric($value)) {
				$return = $g->setInstanceParameter($instance, $type, "spindle_use", $value);
			}
			break;
		case "cpuCount":
			if(is_numeric($value)) {
				$return = $g->setInstanceParameter($instance, $type, "vcpus", $value);
			}
			break;
		case "memoryAmount":
			if(is_numeric($value)) {
				$beparams = array(
					"memory" => $value,
					"minmem" => $value,
					"maxmem" => $value,
				);
				$return = $g->setInstanceParameters($instance, $beparams);
			}
			break;
		}
	break;
	case "hvparam":
		$return = $g->setInstanceParameter($instance, $type, $parameter, $value);
		break;
	}
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/updateNic/:i/:n/:m/:l', function($instance, $nic, $mac, $link) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = $g->setNicParameters($instance, $nic, $mac, $link);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/addNic/:i/:m/:l', function($instance, $mac, $link) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = $g->addNic($instance, $mac, $link);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/growDisk/:i/:d/:a', function($instance, $disk, $amount) use ($app) {
	global $config;
    $g = new ganetiClient($config["rapi-current"]);
    $amount = $amount * 1024;
	$return = $g->growDisk($instance, $disk, $amount);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/addDisk/:i/:s', function($instance, $size) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = $g->addDisk($instance, $size);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/addTag/:i/:s', function($instance, $tag) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = $g->addTag($instance, $tag);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/clusterHvParameter/:t/:p/:v', function($type, $parameter, $value) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = $g->setClusterParameter("hvparams", $type, $parameter, $value);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->post('/clusterIpolicyParameter/:p/:v', function($parameter, $value) use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$return = $g->setClusterParameter("ipolicy", "none", $parameter, $value);
	Header("Content-Type: application/json");
	echo json_encode($return);
	exit;
});

$app->get('/statistics', function() use ($app) {
	global $config;
	$g = new ganetiClient($config["rapi-current"]);
	$cluster = $g->getClusterInfo();
	$clusterName = $g->getConfigName();
	$stats = $g->getStats();

	$app->render('page_statistics.html', array( "config" => $config,
		"clusterName" => $clusterName,
		"vlans" => $config["vlans"][$clusterName],
		"cluster" => $cluster,
		"stats" => $stats,
		));
});


$app->run();

?>
